module/executiondatasync/optimistic_sync/pipeline/pipeline.go

```diff
@@ -0,0 +1,404 @@
+// the state change to children pipelines.
+//
+// Expected error returns during normal operations:
+//   - ErrInvalidTransition: when the transition is invalid
+func (p *Pipeline) transitionTo(newState optimistic_sync.State) error {
+	hasChange, err := p.setState(newState)
+	if err != nil {
+		return err
+	}
+
+	if hasChange {
+		// send notification for all state changes. we require that implementations of [PipelineStateConsumer]
+		// are non-blocking and consume the state updates without noteworthy delay.
+		p.stateConsumer.OnStateUpdated(newState)
+		p.stateChangedNotifier.Notify()
+	}
+
+	return nil
+}
+
+// setState sets the state of the pipeline and logs the transition.
+// Returns true if the state was changed, false otherwise.
+//
+// Expected error returns during normal operations:
+//   - ErrInvalidTransition: when the state transition is invalid
+func (p *Pipeline) setState(newState optimistic_sync.State) (bool, error) {
+	currentState := p.GetState()
+
+	// transitioning to the same state is a no-op
+	if currentState == newState {
+		return false, nil
+	}
+
+	if err := p.validateTransition(currentState, newState); err != nil {
+		return false, fmt.Errorf("failed to transition from %s to %s: %w", currentState, newState, err)
+	}
+
+	if !p.state.CompareAndSwap(int32(currentState), int32(newState)) {
+		// Note: this should never happen since state is only updated within the Run goroutine.
+		return false, fmt.Errorf("failed to transition from %s to %s", currentState, newState)
+	}
+
+	p.log.Debug().
+		Str("old_state", currentState.String()).
+		Str("new_state", newState.String()).
+		Msg("pipeline state transition")
+
+	return true, nil
+}
+
+// validateTransition validates the transition from the current state to the new state.
+//
+// Expected error returns during normal operations:
+//   - ErrInvalidTransition: when the transition is invalid
+func (p *Pipeline) validateTransition(currentState optimistic_sync.State, newState optimistic_sync.State) error {
+	switch newState {
+	case optimistic_sync.StateProcessing:
+		if currentState == optimistic_sync.StatePending {
+			return nil
+		}
+	case optimistic_sync.StateWaitingPersist:
+		if currentState == optimistic_sync.StateProcessing {
+			return nil
+		}
+	case optimistic_sync.StateComplete:
+		if currentState == optimistic_sync.StateWaitingPersist {
+			return nil
+		}
+	case optimistic_sync.StateAbandoned:
+		// Note: it does not make sense to transition to abandoned from persisting or completed since to be in either state:
+		// 1. the parent must be completed
+		// 2. the pipeline's result must be sealed
+		// At that point, there are no conditions that would cause the pipeline be abandoned
+		switch currentState {
+		case optimistic_sync.StatePending, optimistic_sync.StateProcessing, optimistic_sync.StateWaitingPersist:
+			return nil
+		}
+
+	default:
+		return fmt.Errorf("invalid transition to state: %s", newState)
+	}
+
+	return ErrInvalidTransition
+}
```

Comment:
In my head, this would be encapsulated into a concurrency-safe implementation of [`State`](https://github.com/onflow/flow-go/blob/7eeb063bd8630b4500e1750bd4bd2e4b30df921d/module/executiondatasync/optimistic_sync/state.go#L3-L17)